### DISCUSSION TITLE
Graph-Level Features Used for Training not Final Output

### DISCUSSION THREAD

**Discussion by anthonysirico** [pyg-team/pytorch_geometric#8921] [Q&A] [answered]
Has anyone tried using a graph-level feature strictly for the training, and it has nothing to do with the final output? For example, if I'm performing graph classification, where `dataset[0].y` is a 1 or a 0, which is a graph-level feature, can I have another, such as a `float` that aids in training but is not part of the final output, it's still just the 1 or 0.

**Comment by rusty1s** [comment]
You can integrate graph-level features by concatenating it to your input features, e.g.:
```python
x = torch.cat([x, graph_x[batch]], dim=-1)

**Reply by anthonysirico** [reply, marked as answer]
If my `data.x` shape is `[14,2]` and my graph-level feature is `[1]`, how does that work when using `torch.cat`?

**Reply by anthonysirico** [reply]
figured it out, had to ensure the forward pass looked like this:

```
   def forward(self, x, edge_index, batch, graph_features):
        x = F.relu(self.conv1(x, edge_index))
        x = F.relu(self.conv2(x, edge_index))
        x = self.conv3(x, edge_index)
        x = global_mean_pool(x, batch)  # [batch_size, hidden_channels]

        graph_features = graph_features.view(-1, 1)

        x = torch.cat([x, graph_features], dim=-1)

        # 3. Apply a final classifier
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.lin(x)

        return x
```

and `train()` looked like this:

```
    def train(self, loader):
        self.model.train()

        total_loss = 0
        for graphs in loader:
            graphs = graphs.to(self.device)

            out = self.model(graphs.x.to(torch.float), graphs.edge_index, graphs.batch, graphs.graph_features)
            loss = self.criterion(out, graphs.it_y)
            loss.backward()
            self.optimizer.step()
            self.optimizer.zero_grad()
            total_loss += loss.item() * graphs.num_graphs
        return total_loss / len(loader.dataset)
```
